client_query.go

```diff
@@ -1,7 +1,237 @@
+		reader: result,
+	}
+}
+
+func (c *gocbcoreRowReader) NextRow() []byte {
+	return c.reader.NextRow()
+}
+
+func (c *gocbcoreRowReader) MetaData() (*QueryMetadata, error) {
+	metaBytes, err := c.reader.MetaData()
+	if err != nil {
+		return nil, translateGocbcoreError(err)
+	}
+
+	var jsonResp jsonAnalyticsResponse
+
+	err = json.Unmarshal(metaBytes, &jsonResp)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal metadata: %s", err) // nolint: err113, errorlint
+	}
+
+	meta := &QueryMetadata{
+		RequestID: "",
+		Metrics: QueryMetrics{
+			ElapsedTime:      0,
+			ExecutionTime:    0,
+			ResultCount:      0,
+			ResultSize:       0,
+			ProcessedObjects: 0,
+		},
+		Warnings: nil,
+	}
+	meta.fromData(jsonResp)
+
+	return meta, nil
+}
+
+func (c *gocbcoreRowReader) Close() error {
+	err := c.reader.Close()
+	if err != nil {
+		return translateGocbcoreError(err)
+	}
+
+	return nil
+}
+
+func (c *gocbcoreRowReader) Err() error {
+	err := c.reader.Err()
+	if err != nil {
+		return translateGocbcoreError(err)
+	}
+
+	return nil
+}
+
+func translateGocbcoreError(err error) error {
+	var coreErr *gocbcore.ColumnarError
+	if !errors.As(err, &coreErr) {
+		return err
+	}
+
+	if coreErr.HTTPResponseCode == 401 {
+		return newColumnarError(coreErr.Statement, coreErr.Endpoint, coreErr.HTTPResponseCode).
+			withErrorText(coreErr.ErrorText).
+			withCause(ErrInvalidCredential)
+	}
+
+	if len(coreErr.Errors) > 0 {
+		code := int(coreErr.Errors[0].Code)
+
+		descs := make([]columnarErrorDesc, len(coreErr.Errors))
+		for i, desc := range coreErr.Errors {
+			descs[i] = columnarErrorDesc{
+				Code:    desc.Code,
+				Message: desc.Message,
+			}
+		}
+
+		if code == 20000 {
+			return newColumnarError(coreErr.Statement, coreErr.Endpoint, coreErr.HTTPResponseCode).
+				withErrors(descs).
+				withCause(ErrInvalidCredential)
+		}
+
+		if code == 21002 {
+			return newColumnarError(coreErr.Statement, coreErr.Endpoint, coreErr.HTTPResponseCode).
+				withErrors(descs).
+				withCause(ErrTimeout)
+		}
+
+		return newQueryError(coreErr.Statement, coreErr.Endpoint, coreErr.HTTPResponseCode, code, coreErr.Errors[0].Message).
+			withErrors(descs)
+	}
+
+	baseErr := newColumnarError(coreErr.Statement, coreErr.Endpoint, coreErr.HTTPResponseCode).
+		withErrorText(coreErr.ErrorText)
+
+	switch {
+	case errors.Is(coreErr.InnerError, gocbcore.ErrTimeout):
+		baseErr.Cause = ErrTimeout
+		if coreErr.WasNotDispatched {
+			baseErr.errorText = "operation not sent to server, as timeout would be exceeded"
+		}
+	case errors.Is(coreErr.InnerError, context.Canceled):
+		baseErr.Cause = context.Canceled
+		if coreErr.WasNotDispatched {
+			baseErr.errorText = "operation not sent to server, as context was cancelled"
+		}
+	case errors.Is(coreErr.InnerError, context.DeadlineExceeded):
+		baseErr.Cause = context.DeadlineExceeded
+		if coreErr.WasNotDispatched {
+			baseErr.errorText = "operation not sent to server, as context deadline would be exceeded"
```

Comment:
Is the errorText supposed to be the same for DeadlineExceeded and ErrTimeout?

Reply:
No, they're slightly different just to make it clear what happened.

Reply:
Ah MB I read them as the same